src/include/spinlock.h

```diff
@@ -54,6 +54,43 @@ inline void spin_unlock(std::atomic_bool& locked);
 inline void spin_lock(ceph::spinlock& locked);
 inline void spin_unlock(ceph::spinlock& locked);
 
+namespace spin {
+
+template <class MutexT, std::size_t MaxTriesV = 64>
+struct trying_guard final
+{
+  typedef MutexT mutex_type;
+
+  trying_guard(mutex_type& m)
+    : m(m)
+  {
+    if (likely(m.try_lock())) {
+      return;
+    }
+
+    std::size_t tries = MaxTriesV;
+    do {
+      emit_pause();
+    } while (!m.try_lock() && --tries > 0);
+
+    if (!tries) {
+      m.lock();
```

Comment:
I don't understand what's happening here. We ```try_lock()``` a bunch, and then if we fail, we just invoke a normal ```lock()``` anyway? What does this get us in a spinlock?

Reply:
Yes, that's the idea behind the *spinlock with backoff ability*. To throw more light on the concept, let me confront it with the plain spinlock.

The cardinal rule is that a spinlock should be used to protect solely those critical sections that are **really short**. Breaking the law can be punished not only with some wasted cycles, but also with a dangerous starvation driven by positive feedback loop. Try to imagine a situation when there is more OS-level threads for an abstract *requests handling* than CPUs and all of them operate on the same spinlock. Unfortunately, critical section executed by the current lock holder gets prolonged (due to e.g.  preemption). Over time other threads wake, continue but ultimately reach the spinlock and start active waiting. Each new spinner globally increases the pressure on CPU time while being treated with the holder on equal rights. I believe that's the reason why Intel recommends to voluntary yield CPU when a contention handling part of a spinlock became hopeless. Though, plain `sched_yield` doesn't provide an OS with any information about the cause/contented resource. In consequence, the context-switch may happen between two waiters effectively giving nothing.

Nowadays mutexes in Linux are implemented on top of the futex mechanism. This translates into having an atomic inside that can be reached with `try_lock()` (yes, with extra jumps imposed by dynamic linking) to spin for a short while. In glibc-2.23 (used by my machine) `lock()` on a vanilla mutex doesn't behave that way, but instead goes directly to kernel via `futex` syscall if the single `lock cmpxchg` fails.

So, the objection that the `ceph::spin::trying_guard` makes a presumption on how mutex is implemented internally is perfectly valid. Still, I perceive this as lesser evil because:
* it's improbable to encounter a mutex implementation without the user-space augmentation; AFAIK all platforms we do support offer this.
* Having a pthread mutex with the short spinning period would require setting `PTHREAD_MUTEX_ADAPTIVE_NP` on a `std::mutex` instance.

Reply:
Sorry, I've been unclear.

First of all, I understand this (despite the naming) to be entirely spinlocks, not sleeping Mutexes.
So the part that's confusing me is how we're advantaged by implementing a backoff locker as its own type with backoff logic, instead of just relying on the spinlock's own backoff mechanism which we've already written.

Reply:
Well, the naming is really confusing, indeed. `trying_guard` targets mutexes, not spinlocks. It was initially called `ceph::spin::adapting_guard`.

IMHO the back-off mechanism of a sleeping mutex (`std::mutex` + `ceph::spin::trying_guard`) can be more efficient than in the case of pure spinlock. This is because spinlock is entirely user-space thing - OS has no knowledge that can be leveraged during the hopeless part of contention. In contrast to `futex(2)`, `sched_yield` doesn't provide any info about the reason of yielding cpu.